ukvm/ukvm_module_gdb.c

```diff
@@ -361,278 +256,252 @@ unsigned char *getpacket(void)
     }
 }
 
-/* send the packet in buffer.  */
-
-void putpacket(char *buffer)
+/* Send packet without waiting for an ACK from the debugger. */
+static void send_packet_no_ack(unsigned char *buffer)
 {
     unsigned char checksum;
     int count;
-    char ch;
+    unsigned char ch;
 
     /*  $<packet info>#<checksum>.  */
-    do {
-        putDebugChar('$');
-        checksum = 0;
-        count = 0;
-
+    send_char('$');
+    checksum = 0;
+    count = 0;
+
+    ch = buffer[count];
+    while (ch) {
+        send_char(ch);
+        checksum += ch;
+        count += 1;
         ch = buffer[count];
-        while (ch) {
-            putDebugChar(ch);
-            checksum += ch;
-            count += 1;
-            ch = buffer[count];
-        }
-
-        putDebugChar('#');
-        putDebugChar(hexchars[checksum >> 4]);
-        putDebugChar(hexchars[checksum % 16]);
-        flush_debug_buffer();
-    } while (getDebugChar() != '+');
-}
-
-void debug_error(char *format, char *parm)
-{
-    if (remote_debug)
-        fprintf(stderr, format, parm);
-}
-
-
-/* Indicate to caller of mem2hex or hex2mem that there has been an error. */
-static volatile int mem_err;
-
-void set_mem_err(void)
-{
-    mem_err = 1;
-}
-
-/* These are separate functions so that they are so short and sweet
- * that the compiler won't save any registers (if there is a fault to
- * mem_fault, they won't get restored, so there better not be any
- * saved).
- */
-int get_char(char *addr)
-{
-    return *addr;
-}
-
-
-void set_char(char *addr, int val)
-{
-    *addr = val;
-}
-
-
-char *mem2hex(char *mem, char *buf, int count)
-{
-    int i;
-    unsigned char ch;
-
-    for (i = 0; i < count; i++) {
-        ch = get_char(mem++);
-        *buf++ = hexchars[ch >> 4];
-        *buf++ = hexchars[ch % 16];
     }
-    *buf = 0;
-    return buf;
-}
-
 
-/* convert the hex array pointed to by buf into binary to be placed in mem */
-/* return a pointer to the character AFTER the last byte written */
-char *hex2mem(char *buf, char *mem, int count)
-{
-    int i;
-    unsigned char ch;
-
-    for (i = 0; i < count; i++) {
-        ch = hex(*buf++) << 4;
-        ch = ch + hex(*buf++);
-        set_char(mem++, ch);
-    }
-    return mem;
+    send_char('#');
+    send_char(hexchars[checksum >> 4]);
+    send_char(hexchars[checksum % 16]);
 }
 
-
-int gdb_is_pc_breakpointing(uint64_t addr)
+/*
+ * Send a packet and wait for a successful ACK of '+' from the debugger.
+ * An ACK of '-' means that we have to resend.
+ */
+static void send_packet(unsigned char *buffer)
```

Comment:
Any particular reason why `buffer` is an `unsigned char`? This seems to result in a bunch of unnecessary casting all over the place later on, also applies to `packet` and `obuf` in `gdb_handle_exception()` and `send_response()`.